raiden/utils/upgrades.py

```diff
@@ -0,0 +1,60 @@
+import importlib
+import os
+import shutil
+from pathlib import Path
+
+from raiden.exceptions import RaidenDBUpgradeBackupError, RaidenDBUpgradeExecutionError
+from raiden.storage.serialize import JSONSerializer
+from raiden.storage.sqlite import SQLiteStorage
+
+
+class UpgradeManager:
+    """ This class is responsible for figuring out which migrations
+    need to be executed in order to bring the database up to date
+    with the current implementation.
+    """
+
+    MIGRATIONS_MODULE = 'raiden.utils.migrations'
+
+    def __init__(
+            self,
+            db_filename: str,
+            current_version: int,
+            new_version: int,
+    ):
+        self._current_version = current_version
+        self._new_version = new_version
+        self._db_filename = Path(db_filename)
+        self._backup_filename = self._db_filename.parent / Path(
+            f'version{self._current_version}_db.backup',
+        )
+
+    def run(self):
+        storage = SQLiteStorage(str(self._db_filename), JSONSerializer())
+
+        self._backup()
+
+        for version in range(self._current_version, self._new_version):
+            upgrade_module_name = f'v{version}_to_v{version+1}'
+            upgrade_module = self._load_module(upgrade_module_name)
+            upgrade_module.upgrade(storage)
+
+    def restore_backup(self):
+        os.remove(str(self._db_filename))
+        shutil.copy(str(self._backup_filename), str(self._db_filename))
+
+    def _backup(self):
+        shutil.copy(str(self._db_filename), str(self._backup_filename))
```

Comment:
I think so... unless the DB is loaded into memory as soon as it's opened which is highly unlikely. I'll put together a script to make sure.

Reply:
I wanted to actually write a test for this and i found out that i already did:
https://github.com/raiden-network/raiden/pull/3227/files#diff-c1aa4380741c4187017321f7185a2bc3R28

Does this satisfy your concern?

Reply:
No:

> Systems that run automatic backups in the background might try to make a backup copy of an SQLite database file while it is in the middle of a transaction. The backup copy then might contain some old and some new content, and thus be corrupt

> If two processes have open connections to the same database file and one process closes its connection, unlinks the file, then creates a new database file in its place with the same name and reopens the new file, then the two processes will be talking to different database files with the same name. (Note that this is only possible on Posix and Posix-like systems that permit a file to be unlinked while it is still open for reading and writing. Windows does not allow this to occur.) Since rollback journals and WAL files are based on the name of the database file, the two different database files will share the same rollback journal or WAL file. A rollback or recovery for one of the databases might use content from the other database, resulting in corruption. A similar problem occurs if a database file is renamed while it is opened and a new file is created with the old name.

ref: https://www.sqlite.org/howtocorrupt.html

Please, use the backup api https://sqlite.org/backup.html

Reply:
Additionally, I think we should do a `sync` after the backup is made and before starting the upgrade.

Reply:
I went through the link for Sqlite backups and seems like it's a C-API... the only python binding i found (active) was: https://github.com/husio/python-sqlite3-backup

The question is: should we follow this approach? or is it enough knowing that the upgrade manager will run before any DB interaction is done, therefore, it would be quite safe IMO to just copy the file.

Reply:
Nevermind the question above. I tried running Raiden and i figured there was an error in my assumptions. Therefore, i reworked the upgrade manager.